LTE link control: Add support for neighbor cell measurements

jtguggedal opened the pull request:
Adds support and test for neighbor cell measurements:

- APIs to start and stop measurements
- Events to indicate successful or failed measurements
- Test for neighbor cell measurement response parsing

jtguggedal commented:
Keeping as a draft for now pending possible changes to `at_cmd` and/or `at_parser`, and to do more testing. I will also look more into where to allocate the memory for the events, as they can be pretty large in the worst case of 17 neighbor cells, taking up a lot of stack.

@tokangas Could you have a look at the added APIs and event structure for the new events and see what you think?

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-4356
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

jtguggedal commented on include/modem/lte_lc.h:
```diff
@@ -182,6 +188,104 @@ struct lte_lc_cell {
+	 *  95: -46 ≤ RSRP < -45 dBm
+	 *  96: -45 ≤ RSRP < -44 dBm
+	 *  97: -44 ≤ RSRP dBm
+	 *  255: not known or not detectable
```
I'm not sure about these RSRP and RSRQ members. Ideally, we should convert them, but because they are by definition ranges, it's not straight-forward.
If we keep it like this, it needs to be documented in some other way, as this current format doesn't look good in the generated docs.

simensrostad replied:
This is handled the same way in https://github.com/nrfconnect/sdk-nrf/pull/4345. I think the suggested alternatives there was either keep it as is (mapping in API doc) or just refer to the API command specification.

simensrostad commented on lib/lte_link_control/lte_lc_helpers.c:
```diff
@@ -22,6 +22,49 @@
 
 LOG_MODULE_REGISTER(lte_lc_helpers, CONFIG_LTE_LINK_CONTROL_LOG_LEVEL);
 
+static int string_to_int(const char *str_buf, int base, int *output)
```
Should't this function return an unsigned 32-bit integer if in fact a converted cell_id converts to a number larger than an signed 32-bit integer can hold. Lets say the cell_id is DDD11B07 = 3721468679.

simensrostad replied:
Just realized. The maximum E-Cell ID is of course well within a signed 32. So keeping as is should be all good.

jtguggedal replied:
Yep, this should be fine

simensrostad commented on lib/lte_link_control/lte_lc_helpers.c:
```diff
@@ -22,6 +22,49 @@
+	char *end_ptr;
+
+	errno = 0;
+	temp = strtol(str_buf, &end_ptr, base);
```
strtoul should probably be used when cell ids can covert to numbers larger than signed ints.

simensrostad replied:
Nvm😉

simensrostad commented on lib/lte_link_control/lte_lc_helpers.c:
```diff
@@ -22,6 +22,49 @@
+			       size_t idx, int *output, int base)
+{
+	int err;
+	char str_buf[16];
```
If I'm not mistaken this number could be increased to 33. 32 (+ 1 null character) If converting with base 2 to get to the maximum value of a unsigned 32-bit integer. But if exclusively using it for base 16 and 10 conversions 11 should be enough.

jtguggedal replied:
I think I'll leave it as is for now, but do a follow-up PR to use this function in also other places in the helper file. Then the buffer size can be tuned to fit all use-cases

simensrostad commented on lib/lte_link_control/lte_lc_helpers.c:
```diff
@@ -528,3 +570,206 @@ int parse_cereg(const char *at_response,
+
+	if (!response_is_valid(response_prefix, response_prefix_len,
+			       AT_NCELLMEAS_RESPONSE_PREFIX)) {
+		/* The unsolicited response is not a CEREG response, ignore it.
```
```suggestion
		/* The unsolicited response is not a XNCELLMEAS response, ignore it.
```

simensrostad commented on lib/lte_link_control/lte_lc_helpers.c:
```diff
@@ -528,3 +570,206 @@ int parse_cereg(const char *at_response,
+			       AT_NCELLMEAS_RESPONSE_PREFIX)) {
+		/* The unsolicited response is not a CEREG response, ignore it.
+		 */
+		LOG_DBG("Not a valid CEREG response");
```
```suggestion
		LOG_DBG("Not a valid XNCELLMEAS response");
```

simensrostad reviewed (commented):
Just had a pass through. Left a few comments to consider.

jtguggedal commented on lib/lte_link_control/lte_lc.c:
```diff
@@ -314,6 +316,29 @@ static void at_handler(void *context, const char *response)
+		struct lte_lc_ncell neighbor_cells[AT_NCELLMEAS_N_MAX_ARRAY_SIZE];
+		evt.cells_info.neighbor_cells = neighbor_cells;
+
+		if (!evt_handler) {
```
I considered it, but it will not work for the CEREG case which gives a semaphore upon connection establishment also when a handler is not registered

jtguggedal commented on lib/lte_link_control/lte_lc.c:
```diff
@@ -314,6 +316,29 @@ static void at_handler(void *context, const char *response)
+			return;
+		}
+
+		if (parse_ncellmeas(response, &evt.cells_info)) {
```
Agree, this has now been changed. I also added handling of `-E2BIG` in cases where some, but not all neighbor cells were parsed.

jtguggedal commented:
@simensrostad Comments have been addressed, thanks for reviewing. 

I've changed the memory handling quite a bit now to save both heap and stack. I have added some logic to figure out how much memory is actually needed both to parse the AT command and to create the event with the potentially large neighbor cells array. This lowers the space requirements for the heap in most cases where the number of neighboring cell is lower than the maximum 17.

simensrostad reviewed (approved):
LGMT. A few nits.

jtguggedal commented on include/modem/lte_lc.h:
```diff
@@ -185,8 +192,102 @@ struct lte_lc_edrx_cfg {
+	/** Measurement time of serving cell in milliseconds.
+	 *  Range 0 - 18 446 744 073 709 551 614 ms.
+	 */
+	uint32_t measurement_time;
```
It's changed to `uint64_t` now that the `at_params_int64_get` API is merged

jtguggedal commented on lib/lte_link_control/lte_lc_helpers.c:
```diff
@@ -567,3 +623,236 @@ int parse_xt3412(const char *at_response, uint64_t *time)
+
+uint32_t neighborcell_count_get(const char *at_response)
+{
+	uint32_t comma_count, ncell_elements, ncell_count;
```
I usually don't add NULL checks to internal APIs where we have full control of all calls. But added one here, there's no harm in doing it

rlubos commented on lib/lte_link_control/lte_lc_helpers.c:
```diff
@@ -567,3 +622,241 @@ int parse_xt3412(const char *at_response, uint64_t *time)
+	}
+
+	if (status != AT_NCELLMEAS_STATUS_VALUE_SUCCESS) {
+		return 1;
```
Memory leak? it should jump to `clean_exit` to release the list I guess?

jtguggedal replied:
Yeah, that's right.. I even tested for memory leaks this time around, but that didn't help much apparently. Fixed now and added a test case that would have shown the leak.

tokangas reviewed (commented):
Sorry for the delayed review.

tokangas commented on include/modem/lte_lc.h:
```diff
@@ -163,6 +163,12 @@ enum lte_lc_evt_type {
 	LTE_LC_EVT_TAU_PRE_WARNING,
+
+	/** Event containing results from neighbor cell measurements. */
+	LTE_LC_EVT_NEIGHBOR_CELLS,
+
+	/** No cells were found during neightbor cell measurements. */
+	LTE_LC_EVT_NEIGHBOR_CELLS_NONE,
```
Is there a need for a separate event when no neighbor cells are found? In that case `ncells_count` would be zero.

As a generic comment, there's already a lot of events which have associated data in the union. It would be good to say which enum/struct is used with which event.

jtguggedal replied:
> Is there a need for a separate event when no neighbor cells are found? In that case ncells_count would be zero.

This only had a purpose at a point in development, but doesn't really anymore, so I'll remove the event type.

> As a generic comment, there's already a lot of events which have associated data in the union. It would be good to say which enum/struct is used with which event.

Agreed. This is planned to come in a separate PR for all events, together with updated readme and general docs overhaul for the library.

tokangas replied:
> > Is there a need for a separate event when no neighbor cells are found? In that case ncells_count would be zero.

> This only had a purpose at a point in development, but doesn't really anymore, so I'll remove the event type.

Ok, good. The name of the event could also be for example `LTE_LC_EVT_NEIGHBOR_CELL_MEAS`. That way it would refer to the neighbor cell measurement operation. The data would then contain zero or more neighbor cells.

jtguggedal replied:
Updated to the suggested event name, it's better. 
`LTE_LC_EVT_NEIGHBOR_CELL_MEAS` is now also used in the case where measurement status is `1`, meaning no current cell, and in the case where there is a current cell, but no neighbor cell information.  `ncells_count` is set to `0` in both cases and `struct lte_lc_cells_info` documentation is added to clarify that the cell information should only be considerd valid when `ncells_count > 0`.

jtguggedal commented on include/modem/lte_lc.h:
```diff
@@ -489,6 +591,19 @@ int lte_lc_func_mode_get(enum lte_lc_func_mode *mode);
 int lte_lc_lte_mode_get(enum lte_lc_lte_mode *mode);
 
+/**@brief Start neighboring cell measurements. It may take a while
+ *	  before the measurement results are ready,
```
That's right, only one measurement is performed for each time measurements are started. So `_start` and `_stop` might not be a bit misleading.
How about something like this:
```

/**@brief Initiate a neighboring cell measurement.
 *	  The result of the measurement is reported back as an event of the type
 *	  LTE_LC_EVT_NEIGHBOR_CELLS or LTE_LC_EVT_NEIGHBOR_CELLS_NONE, and a
 *	  event handler must be regsitered to reveive the information.
 *	  Depending on the network conditions and LTE connection state, it may
 *	  take a while before the measurement result is ready and reported back.
 *	  After one of the events is received, the neighbor cell measurements
 *	  are automatically stopped.
 *
 * @return Zero on success or (negative) error code otherwise.
 */
int lte_lc_neighbor_cell_measurement(void);

/**@brief Cancel an ongoing neighbor cell measurement.
 *
 * @return Zero on success or (negative) error code otherwise.
 */
int lte_lc_neighbor_cell_measurement_cancel(void);
```

This is closer to AT docs, although the command for stopping the cell measurements is using "stop" instead of cancel.

tokangas replied:
Looks good, I think that makes the interface more clear.

jtguggedal replied:
Updated to the suggestion above, with minor tweaks to docs to reflect removal of `_NONE` event and event name change.

tokangas commented on include/modem/lte_lc.h:
```diff
@@ -445,6 +549,25 @@ int lte_lc_func_mode_get(enum lte_lc_func_mode *mode);
+/**@brief Initiate a neighbor cell measurement.
+ *	  The result of the measurement is reported back as an event of the type
+ *	  LTE_LC_EVT_NEIGHBOR_CELL_MEAS, meaning that an event handler must be
+ *	  regsitered to receive the information.
```
```suggestion
 *	  registered to receive the information.
```

tokangas reviewed (approved):
Thanks, looks good! Only one typo found. :)
